### PR TITLE
Refactor/wrap error type

### LINKED ISSUE #116: Move WrapError to utils
**Is your feature request related to a problem? Please describe.**

- Currently WrapError function lives in `types` which doesn't seem right place
- Interface expects two errors which is not great since, a lot of places has a string input (and error inputs in some place)

**Describe the solution you'd like**
- Move WrapError to utils or similar
- Interface should accept `any` type and then based on type do the Wrapping.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by renlulu**
Fixes #116  .

### CODE REVIEW

**Comment by samlaf** [comment]
@renlulu sorry took us a while to review this, looks good. Just merged another PR which created merge conflict.. really sorry.
Would you mind rebasing on top of master? Then I'll review and we can merge promptly.

**Comment by renlulu** [comment]
> @renlulu sorry took us a while to review this, looks good. Just merged another PR which created merge conflict.. really sorry. Would you mind rebasing on top of master? Then I'll review and we can merge promptly.

Yes resolving!

**Comment by renlulu** [comment]
@samlaf Hey can you confirm the checking error? Looks like there is some permission issue?

**Comment by samlaf** [comment]
> @samlaf Hey can you confirm the checking error? Looks like there is some permission issue?

That's unfortunately expected... the way we upload the badges on the README requires committer to be in eigenlabs org right now.. we have an issue to eventually fix this, but we can merge with this failing.

File: utils/utils.go [hunk 1]
```diff
@@ -256,3 +257,34 @@ func ValidateRawGithubUrl(url string) error {
...
+	if main == nil && sub == nil {
+		return nil
+	}
+
+	if main == nil && sub != nil {
+		return fmt.Errorf("sub error: %w", sub)
+	}
+
+	if main != nil && sub == nil {
+		return fmt.Errorf("%w: unknown sub error", main)
```
Comment (samlaf) [inline utils/utils.go:286]: how about just returning main instead?
this follows the [errors.Join](https://pkg.go.dev/errors#Join) semantic where they just discard any nil error argument

File: utils/utils.go [hunk 2]
```diff
@@ -256,3 +257,34 @@ func ValidateRawGithubUrl(url string) error {
...
+	var main, sub error
+	main = TypedErr(mainErr)
+	sub = TypedErr(subErr)
+	// Some times the wrap will wrap a nil error
+	if main == nil && sub == nil {
+		return nil
+	}
+
+	if main == nil && sub != nil {
+		return fmt.Errorf("sub error: %w", sub)
```
Comment (samlaf) [inline utils/utils.go:282]: is this needed? why not just return sub?

File: utils/utils.go [hunk 3]
```diff
@@ -256,3 +257,34 @@ func ValidateRawGithubUrl(url string) error {
...
+	case string:
+		return errors.New(t)
+	default:
+		return nil
+	}
+}
+
+func WrapError(mainErr interface{}, subErr interface{}) error {
+	var main, sub error
+	main = TypedErr(mainErr)
```
Comment (samlaf) [inline utils/utils.go:274]: can you move all of these to utils/errors.go instead?

**Review by samlaf — CHANGES_REQUESTED** [review]

File: utils/utils.go (hunk 1 above)
Comment (renlulu) [reply utils/utils.go:286]: Okie, make sense. I was just following the initial returned value style. Changing now.

File: utils/utils.go (hunk 3 above)
Comment (renlulu) [reply utils/utils.go]: > Just done!

File: utils/errors.go [hunk 4]
```diff
@@ -23,3 +23,34 @@ var (
...
+
+	if main == nil && sub != nil {
+		return sub
+	}
+
+	if main != nil && sub == nil {
+		return main
+	}
+
+	return main
```
Comment (samlaf) [inline utils/errors.go:55]: you're not wrapping anymore.

**Review by samlaf — CHANGES_REQUESTED** [review]

File: utils/errors.go (hunk 4 above)
Comment (samlaf) [reply utils/errors.go:55]:
```suggestion
	return fmt.Errorf("%w: %w", main, sub)
```

Comment (renlulu) [reply utils/errors.go]: on ok!

**Review by samlaf — APPROVED** [review]
LGTM. reran the flaky test and it passed. Will solve in a separate PR. Let's merge this. :)

**Pull request merged**